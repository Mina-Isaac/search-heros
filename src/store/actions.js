import * as actionTypes from "./constants";

const searchRequest = (query) => ({
  type: actionTypes.SEARCH_REQUEST,
  query,
});

const searchSuccess = (result) => ({
  type: actionTypes.SEARCH_SUCCESS,
  result,
});

const searchFailure = (error) => ({
  type: actionTypes.SEARCH_FAILURE,
  error,
});

export { searchRequest, searchSuccess, searchFailure };
