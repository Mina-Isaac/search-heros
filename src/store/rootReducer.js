import { combineReducers } from "redux";
import loading from "./reducers/loadingReducer";
import result from "./reducers/resultReducer";

export default combineReducers({
  loading,
  result,
});
