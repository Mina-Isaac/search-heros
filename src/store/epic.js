import { of, from } from "rxjs";
import * as actionTypes from "./constants";
import { catchError, map, filter, switchMap } from "rxjs/operators";
import { searchSuccess, searchFailure } from "./actions";

export const fetchSearchResults = (action$, state$, search) =>
  action$.pipe(
    filter((action) => action.type === actionTypes.SEARCH_REQUEST),
    switchMap((action) =>
      from(search(action.query)).pipe(
        map(searchSuccess),
        catchError((error) => of(searchFailure(error)))
      )
    )
  );
