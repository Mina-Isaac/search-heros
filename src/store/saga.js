import { call, put, takeEvery, takeLatest } from "redux-saga/effects";
import * as actions from "./actions";
import { search } from "../api";

function* fetchSearchResult(action) {
  try {
    const result = yield call(search, action.query);
    yield put(actions.searchSuccess(result));
  } catch (err) {
    yield put(actions.searchFailure(err));
  }
}

function* rootSaga() {
  yield takeLatest(actions.searchRequest().type, fetchSearchResult);
}
export default rootSaga;
