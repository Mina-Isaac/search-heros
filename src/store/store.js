import { applyMiddleware, createStore } from "redux";
import createSagaMiddleware from "redux-saga";
import rootReducer from "./rootReducer";
import rootSaga from "./saga";

const sagaMiddleware = createSagaMiddleware();
const enhancer = applyMiddleware(sagaMiddleware);
const store = createStore(rootReducer, enhancer);
sagaMiddleware.run(rootSaga);

export default store;
