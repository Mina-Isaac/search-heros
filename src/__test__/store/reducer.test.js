import reducer from "../../store/rootReducer";
import * as actions from "../../store/actions";
import { heroes as result } from "../../api/db.json";


const getInitialState = (initial) =>
  reducer(initial , {} );

describe("Reducer test", () => {
  it("initial state should match a snapshot", () => {
    const initialState = getInitialState();
    expect(initialState).toMatchSnapshot();
  });

  it("should put retrieved search result in the store and set loading flag to false", () => {
    const initialState = getInitialState();
    const state = reducer(
      initialState,
      actions.searchSuccess(result)
    );
    expect(state.result).toStrictEqual(result);
    expect(state.loading).toEqual(false);
  });

  it("should set state.result to [] and set loading flag to false when an error occures", () => {
    const initialState = getInitialState();
    initialState.loading = true;
    const output = reducer(
      initialState,
      actions.searchFailure(new Error("some error occurred"))
    );
    expect(output.result).toHaveLength(0);
    expect(output.loading).toEqual(false);
  });
});
