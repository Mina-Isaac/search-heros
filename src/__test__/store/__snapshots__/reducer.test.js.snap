// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer test initial state should match a snapshot 1`] = `
Object {
  "loading": false,
  "result": Array [],
}
`;
